server-readonly/src/main/java/org/eclipse/rdf4j/http/server/readonly/QueryResponder.java

```diff
@@ -0,0 +1,190 @@
+			response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+		}
+	}
+
+	private enum QueryTypes {
+		CONSTRUCT_OR_DESCRIBE(q -> q instanceof GraphQuery, RDFFormat.TURTLE, RDFFormat.NTRIPLES, RDFFormat.JSONLD,
+				RDFFormat.RDFXML) {
+			@Override
+			protected void evaluate(Query q, String acceptHeader, HttpServletResponse response)
+					throws QueryEvaluationException, RDFHandlerException, UnsupportedRDFormatException, IOException {
+				GraphQuery gq = (GraphQuery) q;
+				RDFFormat format = (RDFFormat) bestFormat(acceptHeader);
+				gq.evaluate(Rio.createWriter(format, response.getOutputStream()));
+			}
+		},
+		SELECT(q -> q instanceof TupleQuery, TupleQueryResultFormat.JSON, TupleQueryResultFormat.SPARQL,
+				TupleQueryResultFormat.CSV,
+				TupleQueryResultFormat.TSV) {
+			@Override
+			protected void evaluate(Query q, String acceptHeader, HttpServletResponse response)
+					throws QueryEvaluationException, RDFHandlerException, UnsupportedRDFormatException, IOException {
+				TupleQuery tq = (TupleQuery) q;
+				QueryResultFormat format = (QueryResultFormat) bestFormat(acceptHeader);
+				tq.evaluate(QueryResultIO.createTupleWriter(format, response.getOutputStream()));
+			}
+		},
+
+		ASK(q -> q instanceof BooleanQuery, BooleanQueryResultFormat.TEXT, BooleanQueryResultFormat.JSON,
+				BooleanQueryResultFormat.SPARQL) {
+			@Override
+			protected void evaluate(Query q, String acceptHeader, HttpServletResponse response)
+					throws QueryEvaluationException, RDFHandlerException, UnsupportedRDFormatException, IOException {
+				BooleanQuery bq = (BooleanQuery) q;
+				QueryResultFormat format = (QueryResultFormat) bestFormat(acceptHeader);
+				final Optional<BooleanQueryResultWriterFactory> optional = BooleanQueryResultWriterRegistry
+						.getInstance()
+						.get(format);
+				if (optional.isPresent()) {
+					BooleanQueryResultWriter writer = optional.get().getWriter(response.getOutputStream());
+					writer.handleBoolean(bq.evaluate());
+				}
+
+			}
+		};
+
+		private final FileFormat[] formats;
+		private final Predicate<Query> typeChecker;
+
+		QueryTypes(Predicate<Query> typeChecker, FileFormat... formats) {
+			this.typeChecker = typeChecker;
+			this.formats = formats;
+		}
+
+		/**
+		 * Test if the query is of a type that can be answered. And that the accept headers allow for the response to be
+		 * send.
+		 * 
+		 * @param preparedQuery
+		 * @param acceptHeader
+		 * @return true if the query is of the right type and acceptHeaders are acceptable.
+		 * @throws MismatchingAcceptHeaderException
+		 */
+		protected boolean accepts(Query preparedQuery, String acceptHeader) throws MismatchingAcceptHeaderException {
+			if (accepts(preparedQuery)) {
+				if (acceptHeader == null || acceptHeader.isEmpty()) {
+					return true;
+				} else {
+					for (FileFormat format : formats) {
+						for (String mimeType : format.getMIMETypes()) {
+							if (acceptHeader.contains(mimeType))
+								return true;
+						}
+					}
```

Comment:
Instead of hardcoding the accepted formats in the QueryType enum, perhaps we can use a type-based checker again that verifies if the QueryResultWriterRegistry of the required type has a registered writer corresponding with the accept header. 

May be a bit out of scope for the initial setup of this, but it would allow more flexibility in that it would immediately support any sort of custom format that users have come up with.

Reply:
Sounds like a nice improvement. The check itself comes from experience with running sparql.uniprot.org that not everyone asks for sane query/format mixes.